packages/quill/src/core/subscriber.ts

```diff
@@ -0,0 +1,136 @@
+/**
+ * Any object with a named constructor can request an event subscription.
+ */
+interface Source {
+  constructor: { name: string };
+}
+
+/**
+ * A subscription to an event listener with an originating object,
+ * an event target, an event type, a handler function for the event,
+ * and some optional configuration.
+ */
+interface Subscription {
+  source: Source;
+  target: EventTarget;
+  event: string;
+  handler: EventListenerOrEventListenerObject;
+  options?: boolean | AddEventListenerOptions;
+}
+
+const subscribers = new WeakMap<object, Subscriber>();
+
+/**
+ * Creates a Subscriber instance, and binds it to an object.
+ */
+export function createSubscriber(object: Source): Subscriber {
+  const subscriber = new Subscriber();
+  subscribers.set(object, subscriber);
+  return subscriber;
+}
+
+/**
+ * Gets the Subscriber instance bound to the given object.
+ * Throws an error if the binding does not exist.
+ */
+export function getSubscriber(object: Source): Subscriber {
+  const subscriber = subscribers.get(object);
+  if (!subscriber) {
+    throw new Error(
```

Comment:
Should we throw here, or should we just return undefined? The potential to throw means any call of this function would need to be wrapped in a try-catch which doesn't seem ideal

Reply:
I agree, this is not ideal... I also don't like the undefined option though, as we would only trade try/catch blocks for non-null checks. I've got another idea: getting rid of `createSubscriber()` entirely, and creating the new subscriber instance on the first call of `getSubscriber()` for a given object. The drawback is that one might inadvertently create multiple subscribers instead of one if they call `getSubscriber()` with the wrong argument. For this reason, I added some logging on the creation of new subscribers to facilitate debugging. Please let me know what you think :)